cudax/include/cuda/experimental/__container/async_mdarray.cuh

```diff
@@ -0,0 +1,901 @@
+
+  //! @brief Move-assigns a async_mdarray
+  //! @param __other The other async_mdarray.
+  //! Clears the async_mdarray and swaps the contents with \p __other.
+  _CCCL_HIDE_FROM_ABI async_mdarray& operator=(async_mdarray&& __other) noexcept
+  {
+    if (this == _CUDA_VSTD::addressof(__other))
+    {
+      return *this;
+    }
+
+    __buf_     = _CUDA_VSTD::move(__other.__buf_);
+    __mapping_ = _CUDA_VSTD::exchange(__other.__mapping_, mapping_type());
+    __policy_  = _CUDA_VSTD::exchange(__other.__policy_, __policy_t::invalid_execution_policy);
+    return *this;
+  }
+
+  //! @brief Copy-assigns from a different async_mdarray
+  //! @param __other The other async_mdarray.
+  _CCCL_TEMPLATE(class... _OtherProperties)
+  _CCCL_REQUIRES(__properties_match<_OtherProperties...>)
+  _CCCL_HIDE_FROM_ABI async_mdarray& operator=(const async_mdarray<_Tp, _OtherProperties...>& __other)
+  {
+    if (this == reinterpret_cast<const async_mdarray*>(_CUDA_VSTD::addressof(__other)))
+    {
+      return *this;
+    }
+
+    const auto __count = __other.size();
+    if ((__borrow_resource() != __resource_ref_t(__other.__borrow_resource()) || (capacity() != __other.capacity())))
+    {
+      __buffer_t __new_buf{__other.get_memory_resource(), __other.get_stream(), __count};
+      _CUDA_VSTD::swap(__buf_, __new_buf);
+    }
+
+    __mapping_ = __other.__mapping_;
+    this->__copy_cross<const_pointer, __transfer_kind<_OtherProperties...>>(
+      __other.__unwrapped_begin(), __other.__unwrapped_end(), __unwrapped_begin(), __count);
+    __policy_ = __other.__policy_;
+    return *this;
+  }
+
+  //! @brief Move-assigns from a different async_mdarray
+  //! @param __other The other async_mdarray.
+  _CCCL_TEMPLATE(class... _OtherProperties)
+  _CCCL_REQUIRES(__properties_match<_OtherProperties...>)
+  _CCCL_HIDE_FROM_ABI async_mdarray& operator=(async_mdarray<_Tp, _OtherProperties...>&& __other)
+  {
+    if (this == reinterpret_cast<async_mdarray*>(_CUDA_VSTD::addressof(__other)))
+    {
+      return *this;
+    }
+
+    __buf_     = _CUDA_VSTD::move(__other.__buf_);
+    __mapping_ = _CUDA_VSTD::exchange(__other.__mapping_, mapping_type());
+    __policy_  = _CUDA_VSTD::exchange(__other.__policy_, __policy_t::invalid_execution_policy);
+    return *this;
+  }
+
+  //! @brief Assigns an initializer_list to a async_mdarray, replacing its content with that of the initializer_list
+  //! @param __ilist The initializer_list to be assigned
+  _CCCL_HIDE_FROM_ABI async_mdarray& operator=(_CUDA_VSTD::initializer_list<_Tp> __ilist)
+  {
+    const auto __count = __ilist.size();
+    if (capacity() != __count)
+    {
+      __buffer_t __new_buf{get_memory_resource(), get_stream(), __count};
+      _CUDA_VSTD::swap(__buf_, __new_buf);
+    }
+
+    __mapping_ = mapping_type{_CUDA_VSTD::dims<1>{__count}};
+    this->__copy_cross(__ilist.begin(), __ilist.end(), __unwrapped_begin(), __count);
+    return *this;
+  }
+
+  //! @}
+
+  //! @addtogroup iterators
```

Comment:
1. This iterator range does not coincide with the set of elements of the mdarray if the layout mapping is nonexhaustive.

2. There's no single definition of an iterator range if the layout mapping is nonunique.  It depends on whether users want to do read-only access or read-and-write access.

3. The P1684 mdarray design does not include iterators at all.  Therefore, given this and the complexities of defining an iterator range for general layouts, should we consider just not providing `begin` and `end`?